src/core/analytics/visibilityanalyticshandler.js

```diff
@@ -38,15 +46,41 @@ export default class VisibilityAnalyticsHandler {
      */
     if (isIE()) {
       window.addEventListener('unload', () => {
-        if (this._documentVisibilityState !== 'hidden') {
-          this._reportVisibilityChangeEvent('RESULTS_HIDDEN');
+        if (this._previousResultsVisibilityEvent !== RESULTS_VISIBILITY_EVENT.HIDDEN) {
+          this._reportVisibilityChangeEvent(RESULTS_VISIBILITY_EVENT.HIDDEN);
         }
       });
     }
+
+    /**
+     * Page history updates caused by pushState() or replaceState(), such as when a search is performed,
+     * will not trigger a complete page load so the document's visibility state will not change. Use
+     * popstate event and storage listener on QUERY_ID to report result visibility change events for
+     * back/forward page navigation of the same answers page.
+     */
+    window.addEventListener('popstate', () => {
+      const poppedStateQueryId = storage.get(StorageKeys.HISTORY_POP_STATE)?.get('pop-state-queryId');
+      this._analyticsReporterService.setQueryId(poppedStateQueryId);
+      this._previousResultsVisibilityEvent = RESULTS_VISIBILITY_EVENT.HIDDEN;
+      this._reportVisibilityChangeEvent(RESULTS_VISIBILITY_EVENT.HIDDEN);
+    });
+
+    storage.registerListener({
+      eventType: 'update',
+      storageKey: StorageKeys.QUERY_ID,
+      callback: id => {
+        if (this._previousResultsVisibilityEvent !== RESULTS_VISIBILITY_EVENT.UNHIDDEN) {
+          this._previousResultsVisibilityEvent = RESULTS_VISIBILITY_EVENT.UNHIDDEN;
+          this._reportVisibilityChangeEvent(RESULTS_VISIBILITY_EVENT.UNHIDDEN);
+        }
+      }
```

Comment:
Should this check that the `visibilityState` is actually 'visible' before sending the UNHIDDEN event? Also, does this listener fire after the popstate listener?

Reply:
query_id update when there's a search execute and I believe it can be done only if the page is visible so I didn't put in that check. So far it has been visible status when triggered. However, it could be the case that the response takes a while to load and it finish loading in the background when user switch to a different tab. I can include this check! 

I believe this listener will fire after popstate listener if there's a new query executed. However, in storage's popstate function, the call `clearResults` when navigate to a page with no search, would trigger this before the visibilitychangehandler's popstate. But since there's no new query_id, _reportVisibilityChangeEvent wouldn't fire an analytics event for this page.